Refactor tracking of last_written_location to remove it from abstract_objectt

chrisr-diffblue opened the pull request:
Most of the code change is to remove `last_written_location` out of `abstract_objectt`, with a slight extension to the API to have generic read/write functions, which just wrap the existing `read_*` and `write_*` functions provided by some of the current `*_abstract_objectt` classes. There’s also a bit of a change in API so that setting an `abstract_objectt` to TOP now requires calling a `make_top` function (and likewise for clearing TOP, using a `clear_top` function). Other than that checking the logic in the merge functions in `dependency_context_abstract_object` classes, that would be good.

thk123 commented on src/analyses/variable-sensitivity/abstract_enviroment.cpp:
```diff
@@ -72,46 +72,39 @@ abstract_object_pointert abstract_environmentt::eval(
-      sharing_ptrt<union_abstract_objectt> union_abstract_object=
-        std::dynamic_pointer_cast<const union_abstract_objectt>(
-          eval(parent, ns));
+      abstract_object_pointert union_abstract_object=eval(parent, ns);
```
Now you don't need the cast you don't need the cast, I think you can do away with the `is_union` branch (since `parent` is `member_expr.compound()`

thk123 commented on src/analyses/variable-sensitivity/abstract_enviroment.cpp:
```diff
@@ -309,14 +302,11 @@ abstract_object_pointert abstract_environmentt::write(
-    sharing_ptrt<array_abstract_objectt> array_abstract_object=
-      std::dynamic_pointer_cast<const array_abstract_objectt>(lhs);
+    sharing_ptrt<abstract_objectt> array_abstract_object=
+      std::dynamic_pointer_cast<const abstract_objectt>(lhs);
```
Don't think this cast is required?

thk123 commented on src/analyses/variable-sensitivity/abstract_enviroment.cpp:
```diff
@@ -327,36 +317,36 @@ abstract_object_pointert abstract_environmentt::write(
     {
-      sharing_ptrt<union_abstract_objectt> union_abstract_object=
-        std::dynamic_pointer_cast<const union_abstract_objectt>(lhs);
+      sharing_ptrt<abstract_objectt> union_abstract_object=
```
Cast not needed.

thk123 replied:
Also - not sure if the branch here is correct as one passes in the `member_expr` and one passes in `next_member_expr` and I can't work out why they'd be different?

chrisr-diffblue replied:
Good catch - that's an interesting one, and is in the original code before my refactoring, and my refactoring has just made it stand out more. I'll have a think about why one branch would use `member_expr` and the other would use `next_member_expr`....

thk123 commented on src/analyses/variable-sensitivity/abstract_enviroment.cpp:
```diff
@@ -327,36 +317,36 @@ abstract_object_pointert abstract_environmentt::write(
   {
-    sharing_ptrt<pointer_abstract_objectt> pointer_abstract_object=
-      std::dynamic_pointer_cast<const pointer_abstract_objectt>(lhs);
+    sharing_ptrt<abstract_objectt> pointer_abstract_object=
```
No cast required?

thk123 commented on src/analyses/variable-sensitivity/abstract_object.h:
```diff
@@ -114,24 +128,42 @@ class abstract_objectt:public std::enable_shared_from_this<abstract_objectt>
+  virtual locationst get_last_written_locations() const;
+
+  // The one exception is merge in descendant classes, which needs this
+  virtual void make_top() { top=true; }
```
I think these two methods should be `protected`?

chrisr-diffblue replied:
Ideally, yes - but these two methods get called (amongst other places) from `abstract_environment.cpp` (e.g. for assignment to a symbol in `abstract_environmentt::assign`) which is outside of the `abstract_objectt` hierarchy and therefore these have to be public for the time being at least.

thk123 replied:
My concern is that these are supposed to be immutable, and this is used to verify that if the pointer remains the same value, then the value can be assumed to be unchanged. Surely if you want to make something top, you can don this by creating a new top version of the object?

chrisr-diffblue replied:
Sorry - my mistake, I misread the Github layout and thought you were referring to the `get/set_last_written_locations` methods. You are correct, the `make_top`/`clear_top` methods can be made protected (which I've just done).

These two functions are only used to replaces cases where previously the values of top were directly mutated in cases where we basically did `mc=mutable_clone(); mc.top=blah`. I think this is needed for cases where, say, a full array abstract_object is set to TOP, but we don't want to throw away the information about the individual elements of the array.

thk123 commented on src/analyses/variable-sensitivity/array_abstract_object.h:
```diff
@@ -26,6 +26,22 @@ class array_abstract_objectt:public abstract_objectt
+protected:
+  CLONE
+
   virtual abstract_object_pointert read_index(
```
I wonder if these should become private and non-virtual now?

chrisr-diffblue replied:
I think they still need to be protected because in some cases sub-classes of `array_abstract_objectt` explicitly call back to these methods (e.g. doing a write_index on a `constant_array_abstract_objectt` that is bottom, causes that class to explicitly call back to `array_abstract_objectt::write_index`. Making them non-virtual might be ok though.

thk123 commented on src/analyses/variable-sensitivity/dependency_context_abstract_object.cpp:
```diff
@@ -0,0 +1,397 @@
+#include "dependency_context_abstract_object.h"
+#include "full_struct_abstract_object.h"
+
+class ai_baset;
```
How come a forward declaration is sufficient in a cpp file?

chrisr-diffblue replied:
Actually, it wasn't even needed. Left over from earlier work I think.

thk123 commented on src/analyses/variable-sensitivity/dependency_context_abstract_object.h:
```diff
@@ -0,0 +1,159 @@
+
+/// General implementation of an abstract_objectt which tracks the
+/// last-written locations for a given abstract_objectt.
+/// This implementation doesn't need templating, it only delegates
```
What does "this implementation doesn't need templating" mean?

thk123 replied:
Oh I see - there is a templated version that wraps it, you might want to mention that class in the comment

thk123 commented on src/analyses/variable-sensitivity/dependency_context_abstract_object.h:
```diff
@@ -0,0 +1,159 @@
+
+  // To enforce copy-on-write these are private and have read-only accessors
+  abstract_objectt::locationst last_written_locations;
+  abstract_object_pointert ao;
```
Can this have a better name? Perhaps `child_object` or `object_value`?

chrisr-diffblue replied:
I'm about to push a bunch more commits to this PR, but there's no need to re-review just yet as theres still some stuff I want to do. I'll ask for re-review directly once I'm done :-)

thk123 commented on src/analyses/variable-sensitivity/dependency_context_abstract_object.cpp:
```diff
@@ -0,0 +1,397 @@
+
+void dependency_context_abstract_objectt::make_top()
+{
+  abstract_objectt::make_top();
```
Nit: To ensure this method is always called, you could make `make_top` protected and non-virtual, have have a private virtual method `make_top_internal` that it calls for the base classes to override.

thk123 commented on src/analyses/variable-sensitivity/dependency_context_abstract_object.cpp:
```diff
@@ -0,0 +1,397 @@
+
+void dependency_context_abstract_objectt::clear_top()
+{
+  abstract_objectt::clear_top();
```
Ditto for this (protected non-virtual method calling private virtual method)

thk123 commented on src/analyses/variable-sensitivity/dependency_context_abstract_object.cpp:
```diff
@@ -0,0 +1,397 @@
+  return result;
+}
+
+void dependency_context_abstract_objectt::update_sub_elements(
```
What is the purpose of this method? Perhaps it should have a comment. Or perhaps don't overload it since the default implementation does nothing. Or perhaps since it is used for last written location only, maybe the method isn't needed anymore?

chrisr-diffblue replied:
Good catch. The method is still needed in the bigger picture (to enable updates to trickle down into sub-elements of structs/array_abstract_objects), but not actually needed in this particular class and is left over from debugging some stuff.

thk123 commented on src/analyses/variable-sensitivity/dependency_context_abstract_object.cpp:
```diff
@@ -0,0 +1,397 @@
+  abstract_object_pointert updated_child=
+    ao->write(environment, ns, stack, specifier, value, merging_write);
+
+  result->set_child(updated_child);
```
I think here we should check whether the pointer has in fact changed before updating the clone (and if not return this) since we use the fact that a pointer having being modified as an indication that the value has been updated or not (which is then used to find a fixed point)

thk123 replied:
I think comment is addressed by your method to find out if something has changed from a previous version

thk123 commented on src/analyses/variable-sensitivity/dependency_context_abstract_object.cpp:
```diff
@@ -0,0 +1,397 @@
+  // in a more general pattern without also pushing knowledge
+  // of last_written_location housekeeping into the other
+  // *_abstract_objectt classes.
+  if(std::dynamic_pointer_cast<const full_struct_abstract_objectt>
```
Could you extend this comment to cover what this is doing? Is this making so when I do `a.x = 4`, the abstract object for `a` last written location being set to the same as `a.x`?

thk123 replied:
I think it might be preferable to push hose keeping into the other objects as this looks likely to trip over someone implementing a new `struct_abstract_objectt`.

thk123 commented on src/analyses/variable-sensitivity/dependency_context_abstract_object.cpp:
```diff
@@ -0,0 +1,397 @@
+
+  if(cast_other)
+  {
+    bool child_modified;
```
Set default value to prevent undefined behaviour

thk123 commented on src/analyses/variable-sensitivity/dependency_context_abstract_object.cpp:
```diff
@@ -0,0 +1,397 @@
+
+  // Output last written locations immediately following the child output
+  out << " @ ";
+  output_last_written_locations(out, last_written_locations);
```
:+1:

thk123 commented on src/analyses/variable-sensitivity/dependency_context_abstract_object.cpp:
```diff
@@ -0,0 +1,397 @@
+/*******************************************************************\
```
I don't think this file is added to a Makefile?

thk123 reviewed (commented):
For future reference, I think the following could be split into separate commits:

 - The renaming of all the read_* methods to read
 - introduction of `is_top` and `make_top`
 - the introduction of the new class. 

It looks overall good except for the fudge to check the full_struct_abstract_object - though I don't fully understand what it is doing so don't know if there exists a better way.

chrisr-diffblue committed:
Factor out 'is_modified' logic into to the abstract_object implementations

Instead of having the abstract_environment have to know details of the
abstract_objects (e.g. last_written_location) in order to decide what
has been modified between two environments (e.g. at a 3-way merge), make
the individual implementation of abstract_object handle that logic.

src/analyses/variable-sensitivity/dependency_context_abstract_object.cpp

chrisr-diffblue committed:
Add a common interface for read/write operations on abstract_objectt's

chrisr-diffblue committed:
Avoid direct manipulation of abstract_objectt::top

chrisr-diffblue committed:
Factor out last_written_location into it's own abstract_objectt

chrisr-diffblue committed:
Refactor and rename update_last_written_locations to be more generic

Two main changes:
  * update_last_written_locations is renamed to update_location_context
  * the update_sub_element functions/mechanism is replaced with a more
    generically useful visitor mechanism

martin-cs commented on src/analyses/variable-sensitivity/dependency_context_abstract_object.h:
```diff
@@ -0,0 +1,199 @@
+/// context information
+template <class AOT>
+class context_abstract_objectt: public dependency_context_abstract_objectt
+{
```
This makes me mildly nervous but I can accept an answer of "no really, this *is* the easy way".

chrisr-diffblue replied:
:-) In general, doing it this way meant that the abstract_object_factory only had to have very minimal changes. When we come to refactor that as well, then probably there are better ways to do this., but at the moment I'm going to claim 'this is the easy way' unless there is something more specific than just the templating that is making you worried?

martin-cs commented on src/analyses/variable-sensitivity/abstract_enviroment.cpp:
```diff
@@ -327,38 +302,22 @@ abstract_object_pointert abstract_environmentt::write(
-
+      // FIXME: Why does this use next_member when the union branch above
+      // FIXME: uses member_expr?
       const member_exprt next_member_expr(to_member_expr(next_expr));
```
This looks fixed...

thk123 replied:
Does it?

chrisr-diffblue replied:
Ah, actually I can see now that this is actually the same value in both cases, but for some reason is doing a copy here... I'll check to see why that is...

martin-cs reviewed (approved):
This looks a lot better than the previous patch.  I'm happy to include this as part of an experimental release.  To do this I would suggest we don't merge this right now but cherry pick along with a rebase.

thk123 commented on src/analyses/variable-sensitivity/dependency_context_abstract_object.cpp:
```diff
@@ -151,7 +151,12 @@ abstract_object_pointert dependency_context_abstract_objectt::write(
+    std::dynamic_pointer_cast<const dependency_context_abstract_objectt>(value);
+
+  if(value_context)
+    result->set_last_written_locations(
```
Nit: Since statement is spread over multiple lines please use braces

thk123 commented on src/analyses/variable-sensitivity/dependency_context_abstract_object.cpp:
```diff
@@ -345,6 +356,89 @@ abstract_objectt::locationst
+  // For two sets of last written locations to match,
+  // each location in one set must be equal to precisely one location
+  // in the other, since a set can assume at most one match
+  auto before_context=
```
I'd be tempted to move this check before the comment since is another edge case we're dealing with early

thk123 commented on src/analyses/variable-sensitivity/abstract_object.cpp:
```diff
@@ -194,6 +194,61 @@ abstract_object_pointert abstract_objectt::expression_transform(
+   value - the value we are trying to write to the component
+
+ Outputs: The abstract_objectt representing the result of writing
+          to a specific component.
```
I don't think this comment is quite right. Perhaps:

```C++
/*
Outputs: The abstract_objectt representing the result of writing
         the value after applying the remaining stack
*/

thk123 commented on src/analyses/variable-sensitivity/abstract_object.cpp:
```diff
@@ -194,6 +194,61 @@ abstract_object_pointert abstract_objectt::expression_transform(
+   specifier - a modifier expression, such as an array index or field specifier
+          used to indicate access to a specific component
+
+ Outputs: The abstract_objectt representing the value of that component. For
```
Comment needs updating (probably just deleting the bit about component)

thk123 commented on src/analyses/variable-sensitivity/full_struct_abstract_object.h:
```diff
@@ -32,17 +32,16 @@ class full_struct_abstract_objectt:public struct_abstract_objectt
+    const namespacet &ns) const override;
 
-  virtual sharing_ptrt<struct_abstract_objectt> write_component(
+  virtual abstract_object_pointert write(
```
Can you not just not override these and leave the `struct_abstract_object` to forward onto the appropriate virtual methods?

thk123 reviewed (commented):
I've reviewed all but the last commit - which I'll come back to but eyes were beginning to glaze over!

thk123 commented on src/analyses/variable-sensitivity/abstract_object.cpp:
```diff
@@ -420,22 +420,16 @@ Function: abstract_objectt::update_last_written_locations
    An abstract_object_pointer pointing to the cloned, updated object.
 
  Purpose: Creates a mutable clone of the current object, and updates
```
It doesn't create a mutable clone in this case so probably comment should be updated.

thk123 commented on src/analyses/variable-sensitivity/dependency_context_abstract_object.h:
```diff
@@ -62,10 +62,29 @@ class dependency_context_abstract_objectt: public abstract_objectt
+      return element->update_location_context(locations, true);
+    }
+
+    locationst locations;
```
Does this mutate over the course of visitation? Can it be a const reference and/or private?

thk123 commented on src/analyses/variable-sensitivity/abstract_object.cpp:
```diff
@@ -133,16 +133,37 @@ Function: abstract_objectt::abstract_object_merge
+          for a merge.
+
+\*******************************************************************/
+abstract_object_pointert abstract_objectt::abstract_object_merge_internal(
```
You seem to be using a mix of doxygen and old style comments. I think just using doxygen is best since removes the risk of when porting it over to new style anything goes wrong.

thk123 commented on src/analyses/variable-sensitivity/dependency_context_abstract_object.cpp:
```diff
@@ -0,0 +1,391 @@
+  const abstract_object_pointert value,
+  bool merging_write) const
+{
+  // But delegate the write to the child
```
This comment appears to follow on from something but not clear what

thk123 reviewed (approved):
Looks fundamentally fine and a lot cleaner than before! Some suggestions and nitpicks for you to peruse.

martin-cs commented:
This patch set has been incorporated into the latest variable-sensitivity-branch during rebasing.  So this PR isn't relevant any more.  My most profuse thanks to @thk123 for his work on this; the changes have been incorporated and will show up in a future PR.
